12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md

```diff
@@ -0,0 +1,227 @@
+
+    public MyCustomPickerValueConverter(
+        IPublishedSnapshotAccessor publishedSnapshotAccessor,
+        IApiContentRouteBuilder apiContentRouteBuilder)
+    {
+        _publishedSnapshotAccessor = publishedSnapshotAccessor;
+        _apiContentRouteBuilder = apiContentRouteBuilder;
+    }
+
+    public override bool IsConverter(IPublishedPropertyType propertyType)
+        => propertyType.EditorAlias.Equals("My.Custom.Picker");
+
+    public override Type GetPropertyValueType(IPublishedPropertyType propertyType)
+        => typeof(Guid?);
+
+    public PropertyCacheLevel GetDeliveryApiPropertyCacheLevel(IPublishedPropertyType propertyType)
+        => PropertyCacheLevel.Elements;
+
+    public Type GetDeliveryApiPropertyValueType(IPublishedPropertyType propertyType)
+        => typeof(DeliveryApiCustomPicker);
+
+    public object? ConvertIntermediateToDeliveryApiObject(
+        IPublishedElement owner,
+        IPublishedPropertyType propertyType,
+        PropertyCacheLevel referenceCacheLevel,
+        object? inter,
+        bool preview,
+        bool expanding)
+    {
+        if (inter is null)
+        {
+            return null;
+        }
+
+        return BuildDeliveryApiCustomPicker(inter, expanding);
+    }
+
+    private DeliveryApiCustomPicker? BuildDeliveryApiCustomPicker(object inter, bool expanding)
+    {
+        if (!_publishedSnapshotAccessor.TryGetPublishedSnapshot(out IPublishedSnapshot? publishedSnapshot) ||
+            publishedSnapshot?.Content is null)
+        {
+            return null;
+        }
+
+        if (!Guid.TryParse(inter as string, out Guid id))
+        {
+            return null;
+        }
+
+        return new DeliveryApiCustomPicker { Id = id };
+    }
+}
+
+public class DeliveryApiCustomPicker
+{
+    public Guid Id { get; set; }
+
+    public DeliveryApiItemDetails? ItemDetails { get; set; }
+}
+
+public class DeliveryApiItemDetails
+{
+    public string? Name { get; set; }
+
+    public IApiContentRoute? Route { get; set; }
+}
+```
+{% endcode %}
+
+The Implementation of the `IDeliveryApiPropertyValueConverter` interface can be found in the following methods:
+
+* `GetDeliveryApiPropertyCacheLevel()`: This method specifies the cache level used for our property representation in the Delivery API response.
+* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property for the Delivery API response. The value type for the API may differ from the type used for Razor rendering _(as specified by `GetPropertyValueType()`)_.
+  * In the given example, having only the `Guid` as the value is sufficient for the Razor rendering. Since Razor provides full access to the published content within its rendering context, which is not the case with the Content Delivery API.
+* `ConvertIntermediateToDeliveryApiObject()`: This method converts the value from the property editor to the desired custom object in a headless context.
+
```

Comment:
```suggestion

In the given example, the content key (`Guid` value) is used when rendering with Razor. This is sufficient, because Razor provides full access to the published content within the rendering context. However, in a headless context we do not have the same access. To prevent subsequent round-trips to the server, we create a richer output model specifically for the Delivery API.

```

Reply:
Makes a good of sense!